compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/ExpressionCodegen.kt

```diff
@@ -60,110 +52,211 @@ import org.jetbrains.org.objectweb.asm.Type
+        override fun jumpIfTrue(target: Label) = coerced.materialize().also { codegen.mv.ifne(target) }
+        override fun materialize() = coerced.materialize()
+    }
+}
 
-@Suppress("IMPLICIT_CAST_TO_ANY")
 class ExpressionCodegen(
-    val irFunction: IrFunction,
-    val frame: IrFrameMap,
+    private val irFunction: IrFunction,
+    private val frame: IrFrameMap,
     val mv: InstructionAdapter,
     val classCodegen: ClassCodegen
-) : IrElementVisitor<StackValue, BlockInfo>, BaseExpressionCodegen {
+) : IrElementVisitor<PromisedValue, BlockInfo>, BaseExpressionCodegen {
 
-    /*TODO*/
-    val intrinsics = IrIntrinsicMethods(classCodegen.context.irBuiltIns)
+    private val intrinsics = IrIntrinsicMethods(classCodegen.context.irBuiltIns)
 
     val typeMapper = classCodegen.typeMapper
 
-    val returnType = typeMapper.mapReturnType(irFunction.descriptor)
+    val context = classCodegen.context
+
+    private val state = classCodegen.state
+
+    private val fileEntry = classCodegen.context.psiSourceManager.getFileEntry(irFunction.fileParent)
+
+    override val frameMap: IrFrameMap
+        get() = frame
+
+    override val visitor: InstructionAdapter
+        get() = mv
+
+    override val inlineNameGenerator: NameGenerator = NameGenerator("${classCodegen.type.internalName}\$todo") // TODO
+
+    override var lastLineNumber: Int = -1
+
+    private val KotlinType.asmType: Type
+        get() = typeMapper.mapType(this)
+
+    private val IrType.asmType: Type
+        get() = toKotlinType().asmType
 
-    val state = classCodegen.state
+    private val CallableDescriptor.asmType: Type
+        get() = typeMapper.mapType(this)
+
+    val IrExpression.asmType: Type
+        get() = type.asmType
+
+    // Assume this expression's result has already been materialized on the stack
+    // with the correct type.
+    val IrExpression.onStack: MaterialValue
+        get() = MaterialValue(this@ExpressionCodegen, asmType)
 
-    private val sourceManager = classCodegen.context.psiSourceManager
+    val voidValue: MaterialValue
+        get() = MaterialValue(this@ExpressionCodegen, Type.VOID_TYPE)
 
-    private val fileEntry = sourceManager.getFileEntry(irFunction.fileParent)
+    private fun markNewLabel() = Label().apply { mv.visitLabel(this) }
+
+    private fun IrElement.markLineNumber(startOffset: Boolean) {
+        val offset = if (startOffset) this.startOffset else endOffset
+        if (offset < 0) {
+            return
+        }
+        val lineNumber = fileEntry.getLineNumber(offset) + 1
+        assert(lineNumber > 0)
+        if (lastLineNumber != lineNumber) {
+            lastLineNumber = lineNumber
+            mv.visitLineNumber(lineNumber, markNewLabel())
+        }
+    }
+
+    // TODO remove
+    fun gen(expression: IrElement, type: Type, data: BlockInfo): StackValue {
+        expression.accept(this, data).coerce(type).materialize()
```

Comment:
It does look like this pattern (`accept` + `coerce` + `materialize`) is needed in a lot of places though. So I'd consider not removing this function but renaming it (`generateMaterialized` or something shorter) and using everywhere instead